clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -431,56 +431,46 @@ raw2trace_t::merge_and_process_thread_files()
                     if (entry.timestamp.type != OFFLINE_TYPE_TIMESTAMP)
                         FATAL_ERROR("Missing timestamp entry");
                     times[i] = entry.timestamp.usec;
-                    VPRINT(3, "Thread %u timestamp is @0x" ZHEX64_FORMAT_STRING
-                           "\n", (uint)tids[i], times[i]);
+                    VPRINT(3, "File %u timestamp is @0x" ZHEX64_FORMAT_STRING
+                           "\n", i, times[i]);
                 }
                 if (times[i] != 0 && times[i] < min_time) {
                     min_time = times[i];
                     next_tidx = i;
                 }
             }
-            VPRINT(2, "Next thread in timestamp order is %u @0x" ZHEX64_FORMAT_STRING
-                   "\n", (uint)tids[next_tidx], times[next_tidx]);
             tidx = next_tidx;
-            times[tidx] = 0; // Read from file for this thread's next timestamp.
-            size += instru.append_tid(buf, tids[tidx]);
             // We have to write this now before we append any bb entries.
             CHECK((uint)size < MAX_COMBINED_ENTRIES, "Too many entries");
             if (!out_file.write((char*)buf_base, size))
                 FATAL_ERROR("Failed to write to output file");
             buf = buf_base;
             size = 0;
         }
-        VPRINT(4, "About to read thread %d at pos %d\n",
-               (uint)tids[tidx], (int)thread_files[tidx]->tellg());
+        VPRINT(4, "About to read file %d at pos %u\n",
+               tidx, (int)thread_files[tidx]->tellg());
         if (!thread_files[tidx]->read((char*)&in_entry, sizeof(in_entry))) {
             if (thread_files[tidx]->eof()) {
-                // Rather than a FATAL_ERROR we try to continue to provide partial
-                // results in case the disk was full or there was some other issue.
-                WARN("Input file for thread %d is truncated", (uint)tids[tidx]);
-                in_entry.extended.type = OFFLINE_TYPE_EXTENDED;
-                in_entry.extended.ext = OFFLINE_EXT_TYPE_FOOTER;
+                VPRINT(2, "Finish read file %d\n", tidx);
+                times[tidx] = 0;  // Do not read from this file.
+                tidx = (uint)thread_files.size(); // Request thread scan.
+                --file_count;
+                continue;
             } else
-                FATAL_ERROR("Failed to read from file for thread %d", (uint)tids[tidx]);
+                FATAL_ERROR("Failed to read from file %d", tidx);
         }
         if (in_entry.extended.type == OFFLINE_TYPE_EXTENDED) {
             if (in_entry.extended.ext == OFFLINE_EXT_TYPE_FOOTER) {
-                // Push forward to EOF.
-                offline_entry_t entry;
-                if (thread_files[tidx]->read((char*)&entry, sizeof(entry)) ||
-                    !thread_files[tidx]->eof())
-                    FATAL_ERROR("Footer is not the final entry");
```

Comment:
We're now missing two checks:

1) That each thread ends with a footer: this is important to know whether we had thread truncation.  This was checked above at lines 458-462.  I think we have to put this check back in (will have to iterate the threads at the end I guess), along with the warning and workaround.

2) That each file is not truncated.  We probably need a new kind of footer now, a file footer entry.